### PR TITLE
Hid configurator gui

### PR DESCRIPTION
**Pull request by mip1-nordic**
Add simple Gui, extract configurator

### CODE REVIEW

File: scripts/hid_configurator/configurator_core.py [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+            event_data = None
+        else:
+            event_data = data[:data_len]
+
+        return Response(rcpt, event_id, status, event_data)
+
+
+ConfigOption = collections.namedtuple('ConfigOption', 'range event_id help')
+SENSOR_OPTIONS = {
+    'downshift_run':    ConfigOption((10,   2550),   SENSOR_OPT_DOWNSHIFT_RUN,   'Time in milliseconds of switching from mode Run to Rest 1'),
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_core.py]: Maybe we could make descriptions shorter ?
Like : `Time of switching from mode Run to Rest 1 [ms]`
or even: `Switching time: Run -> Rest 1 [ms]`

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    DEVICE_PID, fwinfo, fwreboot, fetch_config, change_config, dfu_transfer
+
+DEVICE_TYPE = {value: key for key, value in DEVICE_PID.items()}
+
+
+class Device:
+    @staticmethod
+    def create_instance(device_type):
+        device = Device()
+        device.device_type = device_type
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/configurator_gui.py]: `device.type`

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        success = change_config(dev, recipient, config_name, config_value, options, module_id)
+        return success
+
+    def fetchcpi(self):
+        config_name = 'cpi'
+        options = SENSOR_OPTIONS
+        module_id = SETUP_MODULE_SENSOR
+        recipient = self.pid
+        dev = self.dev
+        success, val = fetch_config(dev, recipient, config_name, options, module_id)
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/configurator_gui.py]: no check if success ?

File: scripts/hid_configurator/configurator.py [hunk 4]
```diff
@@ -3,129 +3,11 @@
...
-
-
-ConfigOption = collections.namedtuple('ConfigOption', 'range event_id help')
-SENSOR_OPTIONS = {
-    'downshift_run':    ConfigOption((10,   2550),   SENSOR_OPT_DOWNSHIFT_RUN,   'Time in milliseconds of switching from mode Run to Rest 1'),
-    'downshift_rest1':  ConfigOption((320,  81600),  SENSOR_OPT_DOWNSHIFT_REST1, 'Time in milliseconds of switching from mode Rest 1 to Rest 2.'),
-    'downshift_rest2':  ConfigOption((3200, 816000), SENSOR_OPT_DOWNSHIFT_REST2, 'Time in milliseconds of switching from mode Rest 2 to Rest 3.'),
-    'cpi':              ConfigOption((100,  12000),  SENSOR_OPT_CPI,             'CPI resolution of a mouse sensor'),
-}
+from configurator_core import SETUP_MODULE_SENSOR, DEVICE_PID, SENSOR_OPTIONS, get_device_pid, open_device, fwinfo, \
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py [hunk 5]
```diff
@@ -534,7 +89,7 @@ def configurator():
 
     parser = argparse.ArgumentParser()
     parser.add_argument('device_type', help='Selected device type: ' +
-	    ''.join("%s, " % dt for dt in DEVICE_PID)[:-2])
+                                            ''.join("%s, " % dt for dt in DEVICE_PID)[:-2])
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator.py]: PEP 8: continuation line over-indented for visual indent

File: scripts/hid_configurator/configurator.py [hunk 6]
```diff
@@ -548,8 +103,11 @@ def configurator():
     parser_config_sensor_subparsers = parser_config_sensor.add_subparsers(dest='option')
     parser_config_sensor_subparsers.required = True
     for opt_name in SENSOR_OPTIONS.keys():
-        parser_config_sensor_opt = parser_config_sensor_subparsers.add_parser(opt_name, help=SENSOR_OPTIONS[opt_name].help)
-        parser_config_sensor_opt.add_argument('value', help='value range: {} or "fetch" to fetch the value from device'.format(SENSOR_OPTIONS[opt_name].range))
+        parser_config_sensor_opt = parser_config_sensor_subparsers.add_parser(opt_name, help=
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py [hunk 7]
```diff
@@ -565,8 +123,10 @@ def configurator():
     elif args.command == 'config':
         perform_config(dev, args)
 
+
 if __name__ == '__main__':
     try:
         configurator()
     except KeyboardInterrupt:
         pass
+
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator.py]: PEP 8: newline at the end of the file

File: scripts/hid_configurator/configurator_core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/configurator_core.py [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+    try:
+        success, fetched_data = exchange_feature_report(dev, recipient, event_id, None, True)
+    except:
+        success = False
+
+    if success and fetched_data:
+        fmt = '<BIBBHI'
+        assert(struct.calcsize(fmt) <= EVENT_DATA_LEN_MAX)
+
+        (flash_area_id, image_len, ver_major, ver_minor, ver_rev, ver_build_nr) = struct.unpack(fmt, fetched_data)
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/configurator_core.py [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+        return flash_area_id, image_len, ver_major, ver_minor, ver_rev, ver_build_nr
+    else:
+        return None
+
+
+def fwreboot(dev, recipient):
+    event_id = (EVENT_GROUP_DFU << GROUP_FIELD_POS) | (DFU_REBOOT << TYPE_FIELD_POS)
+
+    try:
+        success, fetched_data = exchange_feature_report(dev, recipient, event_id, None, True)
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/configurator_core.py [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+    event_id = (EVENT_GROUP_SETUP << GROUP_FIELD_POS) | (module_id << MOD_FIELD_POS) | (opt_id << OPT_FIELD_POS)
+    logging.debug('Fetch the current value of {} from the firmware'.format(config_name))
+    try:
+        success, fetched_data = exchange_feature_report(dev, recipient, event_id, None, True)
+    except:
+        success = False
+    if success and fetched_data:
+        val = int.from_bytes(fetched_data, byteorder='little')
+
+    return success, val
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/configurator_core.py [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+    else:
+        print('DFU transfer completed')
+        return True
+
+
+def dfu_transfer(dev, recipient, dfu_image, progress_callback):
+    img_length = os.stat(dfu_image).st_size
+    dfu_info = dfu_sync_wait(dev, recipient, False)
+    if not is_ok(dfu_image):
+        return
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+
+import time
+import hid
+
+from configurator_core import get_device_pid, open_device, NORDIC_VID, SENSOR_OPTIONS, SETUP_MODULE_SENSOR, \
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py (hunk 2 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        device.dev = open_device(device_type)
+        return device
+
+    @staticmethod
+    def list_devices():
+        device_list = hid.enumerate(NORDIC_VID, 0)
+
+        for x in device_list:
+            print(x)
+            if x['usage_page'] == 1:
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+                    print('Keyboard')
+                elif x['product_id'] == 21214:
+                    print('Mouse')
+        return device_list
+
+    def perform_fwinfo(self):
+        return fwinfo(self.dev, self.pid)
+
+    def perform_fwreboot(self):
+        fwreboot(self.dev, self.pid)
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+        self.dev.close()
+        time.sleep(9)
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py (hunk 3 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        dev = self.dev
+        success, val = fetch_config(dev, recipient, config_name, options, module_id)
+        return val
+
+    def perform_dfu(self, file, update_progressbar):
+        dfu_image = file
+        return dfu_transfer(self.dev, self.pid, dfu_image, update_progressbar)
+
+
+def test():
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/configurator_gui.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    print("Configurator GUI test")
+    Device.list_devices()
+
+
+if __name__ == '__main__':
+    try:
+        test()
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/configurator_gui.py]: Done

File: scripts/hid_configurator/gui/gui.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+from kivy.config import Config
+from kivy.properties import ObjectProperty
+from kivy.uix.button import Button
+from kivy.uix.floatlayout import FloatLayout
+from kivy.uix.gridlayout import GridLayout
+from kivy.uix.popup import Popup
+
+from configurator_gui import DEVICE_PID, NORDIC_VID, DEVICE_TYPE, Device
+
+Config.set('input', 'mouse', 'mouse,multitouch_on_demand')  # turn off right click red dots
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/gui/gui.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+    def clear_possible_settings(self):
+        print("clear possible settings")
+        self.root.ids.possible_settings_place.clear_widgets()
+        self.root.ids.dfu_buttons_place.clear_widgets()
+
+    def setcpi_callback(self, value):
+        self.device.setcpi(value)
+
+    def fetchcpi_callback(self):
+        returned = self.device.fetchcpi()
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/gui/gui.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    def initialize_dropdown(self):
+        device_list = Device.list_devices()
+        for device in device_list:
+            if device['usage_page'] == GENERIC_DESKTOP_PAGE and device['vendor_id'] == NORDIC_VID:
+                print("add {} to dropdown".format(device['product_id']))
+                self.add_dropdown_button(device['product_id'])
+
+    def initialize_device(self, pid):
+        self.device = Device.create_instance(DEVICE_TYPE[pid])
+        print("initialize device with pid: {}".format(pid))
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/gui/gui.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+
+
+class DfuButtons(GridLayout):
+    def start_dfu_thread(self):
+        app_ref = App.get_running_app()
+        threading.Thread(target=app_ref.dfu, args=(self.update_progressbar,)).start()
+
+    @mainthread
+    def update_progressbar(self, permil):
+        self.ids.pb.value = permil / 10 + 0.1
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: last returned value is 99.9

File: scripts/hid_configurator/gui/gui.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+                                                          image_len,
+                                                          ver_major,
+                                                          ver_minor,
+                                                          ver_rev,
+                                                          ver_build_nr)
+        else:
+            label.text = 'FW info request failed'
+
+    def on_start(self):
+        print('Start Hello')
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/gui/gui.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+
+    def show_load_list(self):
+        content = LoadDialog(load=self.load_list, cancel=self.dismiss_popup)
+        self._popup = Popup(title="Load a file list", content=content, size_hint=(0.9, 0.9))
+        self._popup.open()
+
+    def load_list(self, path, filename):
+        print(path)
+        print(filename)
+        self.filename = filename[0]
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/gui/gui.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+        returned = self.device.fetchcpi()
+        if 100 <= returned <= 12000:
+            return returned
+        else:
+            return 1600
+
+    def initialize_dropdown(self):
+        device_list = Device.list_devices()
+        for device in device_list:
+            if device['usage_page'] == GENERIC_DESKTOP_PAGE and device['vendor_id'] == NORDIC_VID:
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: for now leave it that way

File: scripts/hid_configurator/gui/gui.py [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+
+
+class DfuButtons(GridLayout):
+    def start_dfu_thread(self):
+        app_ref = App.get_running_app()
+        threading.Thread(target=app_ref.dfu, args=(self.update_progressbar,)).start()
+
+    @mainthread
+    def update_progressbar(self, permil):
+        self.ids.pb.value = permil / 10 + 0.1
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: maybe we could sleep here (update only periodically) ?

File: scripts/hid_configurator/gui/gui.py [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+
+
+class Gui(App):
+    def add_mouse_settings(self):
+        mouse_options = MouseOptions()
+        self.root.ids.possible_settings_place.add_widget(mouse_options)
+
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: 2 empty lines between functions

File: scripts/hid_configurator/gui/gui.py [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+    def setcpi_callback(self, value):
+        self.device.setcpi(value)
+
+    def fetchcpi_callback(self):
+        returned = self.device.fetchcpi()
+        if returned:
+            return returned
+        else:
+            return 0
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: return negative value to indicate error

File: scripts/hid_configurator/gui/gui.py [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+            label.text = 'Unrecognised device'
+
+    def add_dropdown_button(self, pid):
+        dropdown = self.root.ids.dropdown
+        button = Button()
+        button.size_hint_y = None
+        button.height = '48dp'
+        button.text = DEVICE_TYPE[pid]
+        button.bind(on_release=lambda btn: self.show_fwinfo())
+        button.bind(on_release=lambda btn: self.show_possible_settings())
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: do we need to add lambda here in every case ?

File: scripts/hid_configurator/gui/gui.py [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+            label.text = 'FW info request failed'
+
+    def show_load_list(self):
+        content = LoadDialog(load=self.load_list, cancel=self.dismiss_popup)
+        self._popup = Popup(title="Load a file list", content=content, size_hint=(0.9, 0.9))
+        self._popup.open()
+
+    def load_list(self, path, files_path):
+        print(path)
+        print(files_path)
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: check if only one file is chosen

File: scripts/hid_configurator/configurator_cli.py [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+            print('{} set to {}'.format(config_name, config_value))
+        else:
+            print('Failed to set {}'.format(config_name))
+
+
+def perform_fwinfo(dev, args):
+    recipient = get_device_pid(args.device_type)
+    info = fwinfo(dev, recipient)
+    if info:
+        flash_area_id, image_len, ver_major, ver_minor, ver_rev, ver_build_nr = info
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/configurator_cli.py]: discussed offline - please use class and __str__ method

[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_cli.py]: Done

File: scripts/hid_configurator/configurator_cli.py [hunk 31]
```diff
@@ -0,0 +1,131 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+
+import argparse
+import logging
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_cli.py]: Discussed offline

File: scripts/hid_configurator/configurator_cli.py [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+    if success:
+        print('Firmware rebooted')
+    else:
+        print('FW reboot request failed')
+
+    return success
+
+
+def configurator():
+    logging.basicConfig(level=logging.ERROR)
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_cli.py]: Discussed offline

File: scripts/hid_configurator/configurator_core.py [hunk 33]
```diff
@@ -51,29 +48,31 @@
 
 NORDIC_VID = 0x1915
 DEVICE_PID = {
-    'desktop_mouse_nrf52832'    : 0x52DA,
-    'desktop_mouse_nrf52810'    : 0x52DB,
-    'gaming_mouse'              : 0x52DE,
-    'keyboard'                  : 0x52DD,
-    'dongle'                    : 0x52DC,
+    'desktop_mouse_nrf52832': 0x52DA,
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/gui/gui.py (hunk 25 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Discussed offline

File: scripts/hid_configurator/gui/gui.py (hunk 26 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: PEP 8

Comment (MarekPieta) [reply scripts/hid_configurator/gui/gui.py]: ok, so let's leave it as it is

File: scripts/hid_configurator/configurator_cli.py [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+    parser_dfu.add_argument('dfu_image', type=str, help='Path to a DFU image')
+    parser_fwinfo = subparsers.add_parser('fwinfo', help='Obtain information about FW image')
+    parser_fwreboot = subparsers.add_parser('fwreboot', help='Request FW reboot')
+    parser_config = subparsers.add_parser('config', help='Write configuration option')
+    config_subparsers = parser_config.add_subparsers(dest='module')
+    config_subparsers.required = True
+    parser_config_sensor = config_subparsers.add_parser('sensor', help='Optical sensor options')
+    parser_config_sensor_subparsers = parser_config_sensor.add_subparsers(dest='option')
+    parser_config_sensor_subparsers.required = True
+    for opt_name in SENSOR_OPTIONS.keys():
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_cli.py]: Done

File: scripts/hid_configurator/gui/gui.py (hunk 27 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: This method is used to get the value for slider.

File: scripts/hid_configurator/gui/gui.py (hunk 29 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: FileChooserListView choose one file by default.

File: scripts/hid_configurator/gui/gui.py (hunk 28 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/configurator_cli.py [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+    for opt_name in SENSOR_OPTIONS:
+        parser_config_sensor_opt = parser_config_sensor_subparsers.add_parser(opt_name,
+                                                                              help=SENSOR_OPTIONS[opt_name].help)
+        parser_config_sensor_opt.add_argument('value',
+                                              help='value range: {} or "fetch" to fetch the value from device'.format(
+                                                  SENSOR_OPTIONS[opt_name].range))
+    args = parser.parse_args()
+
+    dev = open_device(args.device_type)
+    if not dev:
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_cli.py]: `dev is None`

File: scripts/hid_configurator/configurator_core.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/configurator_core.py]: Still changed intents ?

[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/configurator_core.py]: Also in other places

File: scripts/hid_configurator/configurator_core.py [hunk 36]
```diff
@@ -121,36 +120,22 @@ def parse_response(response_raw):
...
-
-
 def create_set_report(recipient, event_id, event_data):
     """ Function creating a report in order to set a specified configuration
         value.
         Recipient is a device product ID. """
 
-    assert(type(recipient) == int)
-    assert(type(event_id) == int)
+    assert (type(recipient) == int)
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/configurator_core.py]: Still space is added

File: scripts/hid_configurator/gui/gui_backend.py [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+    def create_instance(device_type):
+        device = Device()
+        device.device = device_type
+        device.pid = get_device_pid(device_type)
+        device.dev = open_device(device_type)
+        return device
+
+    @staticmethod
+    def list_devices():
+        device_list = hid.enumerate(NORDIC_VID, )
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui_backend.py]: Maybe add second argument

File: scripts/hid_configurator/gui/gui_backend.py [hunk 38]
```diff
@@ -0,0 +1,77 @@
...
+            return False
+
+    def perform_dfu(self, file, update_progressbar):
+        dfu_image = file
+        return dfu_transfer(self.dev, self.pid, dfu_image, update_progressbar)
+
+
+if __name__ == '__main__':
+    try:
+        print("You should use gui.py")
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/gui/gui_backend.py]: Still not fixed

File: scripts/hid_configurator/gui/gui.py [hunk 39]
```diff
@@ -0,0 +1,150 @@
...
+    def dismiss_popup(self):
+        self._popup.dismiss()
+
+    def dfu(self, update_progressbar):
+        dfu_label = self.root.ids.dfu_label
+        dfu_label.text = 'Transfer in progress'
+        print(self.filepath)
+        success = self.device.perform_dfu(self.filepath, update_progressbar)
+        if success:
+            print('Dfu transfer completed')
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: DFU

File: scripts/hid_configurator/gui/gui.py [hunk 40]
```diff
@@ -0,0 +1,150 @@
...
+            label.text = 'there are no available keyboard options for now'
+            self.add_dfu_buttons()
+        elif self.device.pid == DEVICE_PID['gaming_mouse']:
+            label.text = 'mouse options'
+            self.add_mouse_settings()
+            self.add_dfu_buttons()
+        else:
+            label.text = 'Unrecognised device'
+
+    def add_dropdown_button(self, pid):
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: Could we move this to *.kv file?

File: scripts/hid_configurator/configurator_cli.py (hunk 35 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_cli.py]: Done

File: scripts/hid_configurator/configurator_core.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/gui/gui_backend.py (hunk 37 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui_backend.py]: Second argument set to 0

File: scripts/hid_configurator/gui/gui.py (hunk 39 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/gui/gui.py (hunk 40 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: Done

File: scripts/hid_configurator/configurator_core.py [hunk 41]
```diff
@@ -51,29 +48,31 @@
...
+    WRITE_ERROR = 5
     DISCONNECTED_ERROR = 6
-    FAULT              = 99
+    FAULT = 99
+
 
 class Response(object):
     def __init__(self, recipient, event_id, status, data):
         self.recipient = recipient
-        self.event_id  = event_id
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/configurator_core.py]: Still indents are changed ...

File: scripts/hid_configurator/gui/gui_backend.py (hunk 37 above)
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui_backend.py]: add second argument, please

File: scripts/hid_configurator/gui/gui_backend.py [hunk 42]
```diff
@@ -26,7 +26,7 @@ def create_instance(device_type):
 
     @staticmethod
     def list_devices():
-        device_list = hid.enumerate(NORDIC_VID, )
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui_backend.py]: These changes belong to first commit, right ?

File: scripts/hid_configurator/gui/gui.py [hunk 43]
```diff
@@ -128,11 +140,14 @@ def dfu(self, update_progressbar):
         print(self.filepath)
         success = self.device.perform_dfu(self.filepath, update_progressbar)
         if success:
-            print('DFU transfer completed')
-            dfu_label.text = 'DFU transfer completed'
+            info = 'DFU transfer completed'
+            message = 'You can now go back to application and switch firmware'
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui.py]: Maybe `Firmware will be switched on next reboot`?

**Comment by MarekPieta** [comment]
Please fix last commit message: `player` -> `plyer`.
And please rebase the changes.

File: scripts/hid_configurator/gui/gui.py (hunk 43 above)
Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui.py]: ok, so I also change the name of the button from 'Switch to new firmware' to 'Firmware reboot'

File: scripts/hid_configurator/configurator_core.py [hunk 44]
```diff
@@ -130,9 +130,9 @@ def __init__(self, values):
 
     def __str__(self):
         return ('Firmware info\n'
-                '\tFLASH area id: {}\n'
-                '\tImage length: {}\n'
-                '\tVersion: {}.{}.{}.{}').format(self.flash_area_id, self.image_len, self.ver_major, self.ver_minor,
+                '  FLASH area id: {}\n'
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply scripts/hid_configurator/configurator_core.py]: Done

File: scripts/hid_configurator/gui/gui_backend.py [hunk 45]
```diff
@@ -0,0 +1,82 @@
...
+from configurator_core import DEVICE, get_device_type
+from configurator_core import get_device_pid, open_device, fwinfo, fwreboot, fetch_config, change_config, dfu_transfer
+
+GENERIC_DESKTOP_PAGE = 1
+FWREBOOT_TIME = 10
+
+
+class Device:
+    @staticmethod
+    def create_instance(device_type):
```
Comment (MarekPieta) [inline scripts/hid_configurator/gui/gui_backend.py]: `__init__` ?

Comment (mip1-nordic) [reply scripts/hid_configurator/gui/gui_backend.py]: Done

**Pull request closed without merging**